### PR TITLE
[quantization][DRAFT] Disk space consumption improvements for full model quantization

### PR DESCRIPTION
**Pull request by stamalakhov**
This PR fixes population of static `causal_masks`\`position_embeddings` through the layers to save disk space.

It precomputes  static `causal_mask`/`position_embeddings` for using in `llama/quant_decoder_layer` to prevent populating every quantized decoder layer with these statically computed parameters to save disk space.

Using this PR `circle` model for `HuggingFaceTB/SmolLM2-135M-Instruct` is just 105MiB (vs 300 Mib of #492)

Draft: #436 

TICO-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: tico/quantization/wrapq/wrappers/llama/quant_model.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            )
+
+        if position_ids is None:
+            position_ids = cache_position.unsqueeze(0)
+
+        hidden_states = inputs_embeds
+        # create position_embeddings and causal_mask to be shared across the decoder layers
+        causal_mask = self.layers[0].wrapped.get_attention_mask_for(hidden_states)
+        causal_mask = self._fq(causal_mask, self.obs_causal_mask)
+        position_embeddings = self.layers[0].wrapped.get_position_embeddings_for(
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_model.py]: > How about just creating its own mask and embeddings instead of using first layer's? It needs some duplicate codes but can remove a dependency of the first layer.

@mhs4670go 
Ok. I'll fix it.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 2]
```diff
@@ -300,64 +321,14 @@ def fix_inputs(model, tokenizer, input_ids):
...
-            output_attentions: Optional[bool] = None,
-            output_hidden_states: Optional[bool] = None,
-            return_dict: Optional[bool] = None,
-            cache_position: Optional[torch.LongTensor] = None,
-            logits_to_keep: Union[int, torch.Tensor] = 0,
-            **kwargs: Unpack[KwargsForCausalLM],
-        ) -> Union[Tuple, CausalLMOutputWithPast]:
-            # fixed input size, due to position_ids fixed
-            orig_len = input_ids.shape[-1]
-            input_ids = fix_inputs(self, self.tokenizer, input_ids)
```
Comment (mhs4670go) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: Now, `fix_inputs` can be removed.

File: tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py [hunk 3]
```diff
@@ -185,18 +203,17 @@ def forward(
         residual = hidden_states
         hidden_states = self.input_layernorm(hidden_states)
 
-        # to prevent introduction of attention_mask as a parameter let's use preset attention_mask
-        L = hidden_states.size(1)
-        attention_mask = self._slice_causal(L, hidden_states.device)
+        if attention_mask is None or attention_mask.dtype == torch.bool:
```
Comment (mhs4670go) [inline tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py]: Why does this condition come again?

File: tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py [hunk 4]
```diff
@@ -185,18 +203,17 @@ def forward(
...
-            self.rope_sin_template.to(
-                dtype=hidden_states.dtype, device=hidden_states.device
-            ),
-        )
+        if position_embeddings is None:
+            position_embeddings = self.get_position_embeddings_for(hidden_states)
+            cos, sin = position_embeddings
+            position_embeddings = (
+                self._fq(cos.unsqueeze(1), self.obs_cos),
+                self._fq(sin.unsqueeze(1), self.obs_sin),
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py]: @mhs4670go 
Ahh. Actually these lines are introduced to be the same as in `quant_model.py` to make the codes consistent. It occured that transforms that are done inside `quant_attn` produced another constants during tracing. Originally these lines were located at `164, 165` lines of `quant_attn.py` (you can find them in this draft)

```
      cos_u = cos.unsqueeze(unsqueeze_dim)
      sin_u = sin.unsqueeze(unsqueeze_dim)
```
So i summed all of the transforms and applied them once to prevent their populating.

[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py]: @mhs4670go 
The latest version of `quant_attn.py` doesn't use `unsqueeze_dim` , so it can be removed from here also.
